client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,1607 @@
+
+		for node in to_remove {
+			self.on_remove_reserved_peer(node);
+		}
+	}
+
+	/// Change "reserved only" flag. In "reserved only" mode we connect and accept connections to
+	/// reserved nodes only.
+	fn on_set_reserved_only(&mut self, reserved_only: bool) {
+		trace!(target: "peerset", "Set reserved only: {}", reserved_only);
+
+		self.reserved_only = reserved_only;
+
+		if !reserved_only {
+			return self.alloc_slots()
+		}
+
+		// Disconnect all non-reserved peers.
+		self.nodes.keys().for_each(|peer_id| self.drop_connection(*peer_id));
+		self.nodes.clear();
+		self.num_in = 0;
+		self.num_out = 0;
+	}
+
+	/// Get the list of reserved peers.
+	fn on_reserved_peers(&self, pending_response: oneshot::Sender<Vec<PeerId>>) {
+		pending_response.send(self.reserved_nodes.keys().cloned().collect());
+	}
+
+	/// Disconnect the peer.
+	fn on_disconnect_peer(&mut self, peer_id: PeerId) {
+		// Don't do anything if the node is reserved.
+		if self.reserved_nodes.contains_key(&peer_id) {
+			warn!(
+				target: "peerset",
+				"Ignoring request to disconnect reserved peer {} from {:?}.",
+				peer_id, self.set_id,
+			);
+			return
+		}
+
+
+		match self.nodes.remove(&peer_id) {
+			Some(direction) => {
+				trace!(target: "peerset", "Disconnecting peer {} ({:?}).", peer_id, direction);
+				match direction {
+					Direction::Inbound => self.num_in -= 1,
+					Direction::Outbound => self.num_out -= 1,
+				}
+				self.drop_connection(peer_id);
+			},
+			None => {
+				warn!(
+					target: "peerset",
+					"Trying to disconnect unknown peer {} from {:?}.",
+					peer_id, self.set_id,
+				);
+			},
+		}
+	}
+
+	/// Indicate that we received an incoming connection. Must be answered either with
+	/// a corresponding `Accept` or `Reject`, except if we were already connected to this peer.
+	///
+	/// Note that this mechanism is orthogonal to `Connect`/`Drop`. Accepting an incoming
+	/// connection implicitly means `Connect`, but incoming connections aren't cancelled by
+	/// `dropped`.
+	// Implementation note: because of concurrency issues, it is possible that we push a `Connect`
+	// message to the output channel with a `PeerId`, and that `incoming` gets called with the same
+	// `PeerId` before that message has been read by the user. In this situation we must not answer.
+	fn on_incoming_connection(&mut self, peer_id: PeerId, incoming_index: IncomingIndex) {
+		trace!(
+			target: "peerset",
+			"Incoming connection from peer {} ({:?}).",
+			peer_id, incoming_index,
+		);
+
+		if self.reserved_only && !self.reserved_nodes.contains_key(&peer_id) {
+			self.reject_connection(incoming_index);
+			return
+		}
+
+		// Check if the node is reserved first.
+		if let Some(state) = self.reserved_nodes.get_mut(&peer_id) {
+			match state {
+				// If we're already connected, don't answer, as the docs mention.
+				PeerState::Connected(_) => {},
+				PeerState::NotConnected => {
+					// It's questionable whether we should check a reputation of reserved node.
```

Comment:
I don't think it's a bad idea to reject the connection if the peer is banned because there's a reason for why they're banned. The connection will most likely be retried later at which point the reputation has decayed and the connection can be accepted

Reply:
Comment removed.